### PR TITLE
Tests for stats estimation for UNION, EXCEPT and INTERSECT

### PR DESCRIPTION
**Pull request by kokosing**
Tests for stats estimation for UNION, EXCEPT and INTERSECT

### CODE REVIEW

File: presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java [hunk 1]
```diff
@@ -75,21 +75,50 @@ public void testJoin()
...
 
     @Test
-    public void testSetOperations()
+    public void testUnion()
     {
         statisticsAssertion.check("SELECT * FROM nation UNION SELECT * FROM nation",
-                // real count is 25, estimation cannot know all rows are duplicate. TODO create better test for UNION
-                checks -> checks.estimate(OUTPUT_ROW_COUNT, relativeError(0., 1.)));
+                // real count is 25, estimation cannot know all rows are duplicate.
+                checks -> checks.estimate(OUTPUT_ROW_COUNT, relativeError(1, 1)));
```
Comment (findepi) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java:82]: i would just remove this test case, what do you think?

File: presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java [hunk 2]
```diff
@@ -75,21 +75,50 @@ public void testJoin()
...
+    public void testIntersect()
+    {
         statisticsAssertion.check("SELECT * FROM nation INTERSECT SELECT * FROM nation",
-                checks -> checks.estimate(OUTPUT_ROW_COUNT, relativeError(0, 1.)));
+                // real count is 25, estimation cannot know all rows are duplicate.
+                checks -> checks.estimate(OUTPUT_ROW_COUNT, relativeError(.7, .9)));
+
+        statisticsAssertion.check("SELECT * FROM orders WHERE o_custkey < 900 INTERSECT SELECT * FROM orders WHERE o_custkey > 600",
+                // TODO fix INTERSECT stats calculation as custkey values distirbution is pretty linear
+                checks -> checks.estimate(OUTPUT_ROW_COUNT, relativeError(4, 5)));
```
Comment (findepi) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java:109]: That's a lot!

Also, Maybe it would be worthy to add the tests to test-local-stats too so that we know how this runs without exchanges -- maybe the transformations resulting in query plan that is hard to estimate are run _after_ reorder joins would run?

**Review by findepi — COMMENTED** [review]
quick pass only

File: presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java (hunk 1 above)
Comment (kokosing) [reply presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java:82]: I would keep it. It is corner case which it is nice to know about.

File: presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java (hunk 2 above)
Comment (kokosing) [reply presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java:109]: It is transformed before add exchanges, then add exchanges and add local exchanges transform union to exchange.

**Review by kokosing — COMMENTED** [review]
Addressed comments, please merge it somewhere. I would like to investigate the issues marked with added TODOs

**Comment by kokosing** [comment]
I checked tests marked with TODO and they are estimated ok. Currently, with a single range it is difficult to do anything better.

File: presto-main/src/main/java/com/facebook/presto/cost/PlanStatsAndCostsPrinter.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+        private static String indentString(int indent)
+        {
+            return Strings.repeat("  ", indent);
+        }
+
+        static final class Context
+        {
+            private final int indent;
+            private final StringBuilder output;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/cost/PlanStatsAndCostsPrinter.java:119]: `output` is always the same, so it's more an attribute of the visitor than a visiting context.

**Comment by findepi** [comment]
I cherry-picked `Tests for stats estimation for UNION, EXCEPT and INTERSECT` (as a fixup) into the CBO branch. I don't know what to do with the other commits.
Maybe for now, keep them on a branch so that they can be easily applied to any code version, as needed for development?

**Comment by kokosing** [comment]
Extracted https://github.com/starburstdata/presto/pull/26
